### PR TITLE
MB-35347: Synonym Search

### PR DESCRIPTION
**Pull request by CascadingRadium**
- Allow setting up `synonym_sources` in the index mapping, which will follow its own ingest pipeline, ingesting special synonym definitions using the IndexSynonym API().
- A `synonym_source` can be set like an analyzer to a field mapping and can be set as a default option at the document mapping or the index mapping level.
- Each `synonym_source` can have its own analyzer, making it flexible to allow for compatibility with the language analyzer specified for its corresponding mapping.
- Compatibility with every term-based query where the term gets expanded to include its synonyms at query time.
- Dependencies:
    - blevesearch/bleve_index_api@v1.2.0 - https://github.com/blevesearch/bleve_index_api/pull/57
    - blevesearch/scorch_segment_api@v2.3.0 - https://github.com/blevesearch/scorch_segment_api/pull/46
    - blevesearch/vellum@v1.1.0 - https://github.com/blevesearch/vellum/pull/22
    - blevesearch/zapx@v16@latest - https://github.com/blevesearch/zapx/pull/268

### CODE REVIEW

File: index_impl.go [hunk 1]
```diff
@@ -449,12 +450,25 @@ func (i *indexImpl) preSearch(ctx context.Context, req *SearchRequest, reader in
 		}
 	}
 
+	var fts search.FieldTermSynonymMap
```
Comment (metonymic-smokey) [inline index_impl.go:453]: Please change the name from "fts":)

File: search/util.go [hunk 2]
```diff
@@ -144,5 +145,23 @@ type ScoreExplCorrectionCallbackFunc func(queryMatch *DocumentMatch, knnMatch *D
...
+
+func (f *FieldTermSynonymMap) MergeWith(fts FieldTermSynonymMap) {
+	for field, termSynonymMap := range fts {
+		// Ensure the field exists in the receiver
+		if _, exists := (*f)[field]; !exists {
+			(*f)[field] = make(map[string][]string)
+		}
+		for term, synonyms := range termSynonymMap {
+			// Append synonyms
+			(*f)[field][term] = append((*f)[field][term], synonyms...)
```
Comment (abhinavdangeti) [inline search/util.go:159]: Doesn't look very GO idiomatic at all, with the above change this can become `f[field][term]` ?

File: search/util.go [hunk 3]
```diff
@@ -144,5 +145,23 @@ type ScoreExplCorrectionCallbackFunc func(queryMatch *DocumentMatch, knnMatch *D
 type SearcherStartCallbackFn func(size uint64) error
 type SearcherEndCallbackFn func(size uint64) error
 
+// field -> term -> synonyms
+type FieldTermSynonymMap map[string]map[string][]string
+
+func (f *FieldTermSynonymMap) MergeWith(fts FieldTermSynonymMap) {
```
Comment (abhinavdangeti) [inline search/util.go:151]: Can this not be ..

```
func (f FieldTermSynonymMap) MergeWith(..)
```

File: mapping/synonym.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+import (
+	"fmt"
+
+	"github.com/blevesearch/bleve/v2/analysis"
+	"github.com/blevesearch/bleve/v2/registry"
+)
+
+type SynonymSource struct {
+	CollectionName string `json:"collection"`
```
Comment (abhinavdangeti) [inline mapping/synonym.go:25]: There is no notion of collection for a library user - I suppose you'll be adding a separate documentation on how the library users can benefit from.

File: analysis/type.go [hunk 5]
```diff
@@ -106,6 +106,15 @@ type DateTimeParser interface {
 	ParseDateTime(string) (time.Time, string, error)
 }
 
+const SynonymSourceType = "synonym"
+
+type SynonymSourceVisitor func(name string, item SynonymSource) error
+
+type SynonymSource interface {
+	Analyzer() string
+	Collection() string
```
Comment (abhinavdangeti) [inline analysis/type.go:115]: How does `Collection()` as a source translate to an independent library user?

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Review by abhinavdangeti — APPROVED** [review]

**Comment by abhinavdangeti** [comment]
Thanks @CascadingRadium . Let's get this in for now and incrementally improve the area as and when needed.

**Comment by CascadingRadium** [comment]
Thank you for merging @abhinavdangeti

**Pull request merged**